Compare DDP static graph (C++ core) with legacy DDP forward and backward delay.

bowangbj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61507**

Summary:

Benchmark Python-only DDP vs production C++ based DistributedDataParallel.
- Implemented a pure python DDP: PythonDDP with support of SYNC and ASYNC reduction
- Added compare_ddp to measure the difference in forward and backward step

Kudos on Shen and Yi for the great idea.

Test Plan:

Test on DevGPUS with 2 CUDA devices.

$python compare_ddp.py

Python only DDP has slightly better (-1%) forward performance and slightly slower (2%-20%) backward performance.
This suggested that we need to keep C++ Core since the maximum latency increase can be 20%. See README.md for details.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29685364](https://our.internmc.facebook.com/intern/diff/D29685364)

bowangbj committed:
DDP Benchmark
- Part 1: report forward and backward performace for C++ based DistributedDataParallel
- Part 2: Copy Research code with Pure Python implementation
- Part 3: Implement Pure Python DDP

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

bowangbj committed:
add PythonDDP with BuildBucket support on "DDP Benchmark"


- Part 1: report forward and backward performace for C++ based DistributedDataParallel
- Part 2: Copy Research code with Pure Python implementation
- Part 3: Implement Pure Python DDP
	Bucket, BuildBucketForParam, Add Grad Hooks

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

bowangbj committed:
add PythonDDP with BuildBucket support on "DDP Benchmark"


- Part 1: report forward and backward performace for C++ based DistributedDataParallel
- Part 2: Copy Research code with Pure Python implementation
- Part 3: Implement Pure Python DDP
	Bucket, BuildBucketForParam, Add Grad Hooks

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

bowangbj committed:
PythonDDP AllReduce SYNC works on "DDP Benchmark"


- Part 1: report forward and backward performace for C++ based DistributedDataParallel
- Part 2: Copy Research code with Pure Python implementation
- Part 3: Implement Pure Python DDP
	Bucket, BuildBucketForParam, Add Grad Hooks

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

bowangbj committed:
Add support for sync and async reduction on "DDP Benchmark"


- Part 1: report forward and backward performace for C++ based DistributedDataParallel
- Part 2: Copy Research code with Pure Python implementation
- Part 3: Implement Pure Python DDP which supports sync reduction and async reduction

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29685364](https://our.internmc.facebook.com/intern/diff/D29685364)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

bowangbj committed:
DDP benchmark ready for review on "Benchmark Python-only DDP vs production C++ based DistributedDataParallel."


- Implement a pure python DDP: LegacyDDP, add support for sync and async reduction
- Port LegacyDistributedDataParallel (from fair) as reference
- Write a tool compare_ddp to measure the difference in forward and backward step

Summary:

Test Plan:

$python compare_ddp.py

Python only DDP has slightly better forward performance and slightly slower backward performance.
This validates the idea that a pure python impl of DDP is feasible. See compare_ddp_sample.md for details.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29685364](https://our.internmc.facebook.com/intern/diff/D29685364)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

bowangbj committed:
update summary on "Compare DDP static graph (C++ core) with legacy DDP forward and backward delay."


Summary:

Benchmark Python-only DDP vs production C++ based DistributedDataParallel.
- Implement a pure python DDP: LegacyDDP, add support for sync and async reduction
- Port LegacyDistributedDataParallel (from fair) as reference
- Write a tool compare_ddp to measure the difference in forward and backward step

Kudos on Shen and Yi for the great idea.

Test Plan:

$python compare_ddp.py

Python only DDP has slightly better forward performance and slightly slower backward performance.
This validates the idea that a pure python impl of DDP is feasible. See compare_ddp_sample.md for details.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29685364](https://our.internmc.facebook.com/intern/diff/D29685364)

[ghstack-poisoned]

bowangbj committed:
update sample run on "Compare DDP static graph (C++ core) with legacy DDP forward and backward delay."


Summary:

Benchmark Python-only DDP vs production C++ based DistributedDataParallel.
- Implement a pure python DDP: LegacyDDP, add support for sync and async reduction
- Port LegacyDistributedDataParallel (from fair) as reference
- Write a tool compare_ddp to measure the difference in forward and backward step

Kudos on Shen and Yi for the great idea.

Test Plan:

Test on DevGPUS with 2 CUDA devices.

$python compare_ddp.py

Python only DDP has slightly better forward performance and slightly slower backward performance.
This validates the idea that a pure python impl of DDP is feasible. See compare_ddp_sample.md for details.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29685364](https://our.internmc.facebook.com/intern/diff/D29685364)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

bowangbj committed:
fix lint on "Compare DDP static graph (C++ core) with legacy DDP forward and backward delay."


Summary:

Benchmark Python-only DDP vs production C++ based DistributedDataParallel.
- Implement a pure python DDP: LegacyDDP, add support for sync and async reduction
- Port LegacyDistributedDataParallel (from fair) as reference
- Write a tool compare_ddp to measure the difference in forward and backward step

Kudos on Shen and Yi for the great idea.

Test Plan:

Test on DevGPUS with 2 CUDA devices.

$python compare_ddp.py

Python only DDP has slightly better forward performance and slightly slower backward performance.
This validates the idea that a pure python impl of DDP is feasible. See compare_ddp_sample.md for details.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29685364](https://our.internmc.facebook.com/intern/diff/D29685364)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

bowangbj commented:
Hey Shen, Yi 

The PR is ready for review. I've added two Python-only impls (Legacy DistributedDataParallel from FAIR, pure Python DDP which support async and sync reduction) and compared the fwd and bwd delays, pls see PR summary and compare_ddp_sample.md for test output.  

Key findings: 
1) Good news the Python only Async reduction looks 'slightly' slower than production DDP (as expected). 
2) Async version is much faster than sync version. My impl is similar as legacy impl from FAIR. 

Note that keeping the legacy impl from FAIR here just for reference, pls let me know whether we need to exclude it from this PR.  Looking forward to your cmts.

CC-ed Jerry as FYI.

mrshenli commented on benchmarks/distributed/ddp/compare_ddp.py:
```diff
@@ -0,0 +1,150 @@
+
+        # Backward
+        start.record()
+        loss_fn(outputs, labels).backward()
```
nit: `loss_fn()` should be part of forward.

bowangbj replied:
Done.  maybe in distributed world, backward is more like "reduction" across processes ?

bowangbj replied:
After a second thought, it might still make sense to have loss_func.backward in the backward step. 

Otherwise, it seems there will be nothing to measure in the backward step for production DDP?  Just found that the delay for DDP backward is < 1ms since it has nothing to do if loss func is not included :)

mrshenli replied:
oh, this is surprising and doesn't look right. We previously measured things like below, and usually backward is 3x-5x slower than the forward, and it will be even slower when there are communications. See Fig. 6 in http://www.vldb.org/pvldb/vol13/p3005-li.pdf

```python
fwd_start.record()
loss = loss_fn(model(inputs), labels)
fwd_stop.record()

bwd_start.record()
loss.backward()
bwd_stop.record()
```

If you consistently see this behavior (backward delay is negligible compared to forward), could you please share a small repro?

bowangbj replied:
Sry I misread your cmt. 

I misplaced 
loss_fn(outputs, labels).backward()   -- in forward step

forward: 
outputs = model(inputs)
loss_fn(outputs, labels).backward()

backward:
nothing    :)

Changed it back to sth like 
forward:
outputs = models(inputs)
loss = loss_fn(labels, outputs)

backward:
loss.backward()

Tried it a few times, moving loss_fn(labels, outputs) would not make a big difference as expected.

mrshenli commented on benchmarks/distributed/ddp/compare_ddp.py:
```diff
@@ -0,0 +1,150 @@
+        # for all_reduce does not exist yet.
+        if ddp_option != DDPOption.DDP_CPP_CORE:
+            ddp_model.all_reduce_grads()
+        optimizer.step()
+        optimizer.zero_grad()
```
Let's exclude optimizer from the backward pass

bowangbj replied:
Done

mrshenli commented on benchmarks/distributed/ddp/compare_ddp.py:
```diff
@@ -0,0 +1,150 @@
+        print(f'\nMetrics for GPU {rank} ddp_option {ddp_option}:')
+        for step, elapsed_milliseconds in metrics.items():
+            A = np.array(elapsed_milliseconds)
+            print(' {N} iterations, {event}, mean={mean} ms, median={median} ms, p90={p90} ms, p99={p99} ms'.format(
```
Shall we exclude the 1st 20 iterations to get rid of CUDACachingAllocator warmup overhead?

bowangbj replied:
Great point. Done.

mrshenli commented on benchmarks/distributed/ddp/compare_ddp_sample.md:
````diff
@@ -0,0 +1,23 @@
+- async PythonDDP is slightly slower than DDPCore. DDP_CPP_CORE > PYTHON_DDP_ASYNC > PYTHON_DDP_SYNC > LEGACY_DISTRIBUTED_DATA_PARALLEL
+
+```
+Metrics for GPU 0 ddp_option DDPOption.DDP_CPP_CORE:
````
Hey @bowangbj, thanks a lot for putting this comprehensive measurements together within a very short period of time! And the result looks very promising! Shall we try varying the bucket size for both C++ DDP and Python DDP and see how that affects the delay? In theory, this shouldn't make too much difference, as it does not affects the number of C++/Python context switches. But it will be helpful to know that for sure. :)

bowangbj replied:
Great suggestion! 

Added the result per different bucket size. Overall Python_DDP is 2-20% slower than DDPCore for different bucket size. As expected, larger bucket size can reduce parallelization thus it introduces small delay as bucket size increases.  Also once the bucket size is quite large, it won't make difference as bucket size continues to increase.

Please see details at sample_output.md

mrshenli commented:
> Note that keeping the legacy impl from FAIR here just for reference, pls let me know whether we need to exclude it from this PR. Looking forward to your cmts.

Since your Python DDP implementation is much faster than the legacy one, I'd vote for excluding the legacy one from this PR.

mrshenli reviewed (commented):
nit: shall we put the new files into their own folder? (say benchmarks/distributed/ddp/compare) Asking because `benchmarks/distributed/ddp` already has a README.md, which is irrelevant to the benchmark added in this PR.

bowangbj committed:
resolve shen cmt on "Compare DDP static graph (C++ core) with legacy DDP forward and backward delay."


Summary:

Benchmark Python-only DDP vs production C++ based DistributedDataParallel.
- Implemented a pure python DDP: PythonDDP with support of SYNC and ASYNC reduction
- Added compare_ddp to measure the difference in forward and backward step

Kudos on Shen and Yi for the great idea.

Test Plan:

Test on DevGPUS with 2 CUDA devices.

$python compare_ddp.py

Python only DDP has slightly better (-1%) forward performance and slightly slower (2%-20%) backward performance.
This validates the idea that a pure python impl of DDP is feasible. See README.md for details.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29685364](https://our.internmc.facebook.com/intern/diff/D29685364)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

bowangbj commented:
> > Note that keeping the legacy impl from FAIR here just for reference, pls let me know whether we need to exclude it from this PR. Looking forward to your cmts.
> 
> Since your Python DDP implementation is much faster than the legacy one, I'd vote for excluding the legacy one from this PR.

SG, removed legacy impl

bowangbj commented:
> nit: shall we put the new files into their own folder? (say benchmarks/distributed/ddp/compare) Asking because `benchmarks/distributed/ddp` already has a README.md, which is irrelevant to the benchmark added in this PR.

SG, moved to compare dir

bowangbj committed:
quote summary on "Compare DDP static graph (C++ core) with legacy DDP forward and backward delay."


Summary:

Benchmark Python-only DDP vs production C++ based DistributedDataParallel.
- Implemented a pure python DDP: PythonDDP with support of SYNC and ASYNC reduction
- Added compare_ddp to measure the difference in forward and backward step

Kudos on Shen and Yi for the great idea.

Test Plan:

Test on DevGPUS with 2 CUDA devices.

$python compare_ddp.py

Python only DDP has slightly better (-1%) forward performance and slightly slower (2%-20%) backward performance.
This validates the idea that a pure python impl of DDP is feasible. See README.md for details.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29685364](https://our.internmc.facebook.com/intern/diff/D29685364)

[ghstack-poisoned]

bowangbj commented:
Thx Shen for the super quick and great review. Resolved all the cmts, PTAL

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

mrshenli commented on benchmarks/distributed/ddp/compare/compare_ddp.py:
```diff
@@ -0,0 +1,254 @@
+    PYTHON_DDP_SYNC_REDUCTION = 2
+    PYTHON_DDP_ASYNC_REDUCTION = 3
+
+class LatencyData:
```
Shall we fix this lint error?

```
[B903] Data class should either be immutable or use __slots__ to save memory. Use collections.namedtuple to generate an immutable class, or enumerate the attributes in a __slot__ declaration in the class to leave attributes mutable.
```

bowangbj replied:
SG,  updated with __slots__

mrshenli commented on benchmarks/distributed/ddp/compare/compare_ddp.py:
```diff
@@ -0,0 +1,254 @@
+                data.append(row)
+
+            # Output buffer_size, step as a table.
+            print (tabulate(data,
```
Shall we fix lint errors in this file as well?

bowangbj replied:
Done

mrshenli commented on benchmarks/distributed/ddp/compare/python_ddp.py:
```diff
@@ -0,0 +1,228 @@
+    synchronization and reduction. The only C++ dependency is the common utils:
+    ``dist.all_reduce``
+
+    The idea: parallize gradient calculation and reduction, the same algo as
```
parallize -> parallelize

bowangbj replied:
Great catch

mrshenli reviewed (approved):
LGTM! Thanks for conducting comprehensive experiments! These experiments suggested that we should keep the C++ core, as 20% speed regression is too much.

mrshenli commented:
Please fix all lint errors before landing. Thanks!

bowangbj committed:
fix lint on "Compare DDP static graph (C++ core) with legacy DDP forward and backward delay."


Summary:

Benchmark Python-only DDP vs production C++ based DistributedDataParallel.
- Implemented a pure python DDP: PythonDDP with support of SYNC and ASYNC reduction
- Added compare_ddp to measure the difference in forward and backward step

Kudos on Shen and Yi for the great idea.

Test Plan:

Test on DevGPUS with 2 CUDA devices.

$python compare_ddp.py

Python only DDP has slightly better (-1%) forward performance and slightly slower (2%-20%) backward performance.
This suggested that we need to keep C++ Core since the maximum latency increase can be 20%. See README.md for details.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29685364](https://our.internmc.facebook.com/intern/diff/D29685364)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

bowangbj committed:
reset default epochs to 120 on "Compare DDP static graph (C++ core) with legacy DDP forward and backward delay."


Summary:

Benchmark Python-only DDP vs production C++ based DistributedDataParallel.
- Implemented a pure python DDP: PythonDDP with support of SYNC and ASYNC reduction
- Added compare_ddp to measure the difference in forward and backward step

Kudos on Shen and Yi for the great idea.

Test Plan:

Test on DevGPUS with 2 CUDA devices.

$python compare_ddp.py

Python only DDP has slightly better (-1%) forward performance and slightly slower (2%-20%) backward performance.
This suggested that we need to keep C++ Core since the maximum latency increase can be 20%. See README.md for details.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29685364](https://our.internmc.facebook.com/intern/diff/D29685364)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

bowangbj committed:
another lint fix on "Compare DDP static graph (C++ core) with legacy DDP forward and backward delay."


Summary:

Benchmark Python-only DDP vs production C++ based DistributedDataParallel.
- Implemented a pure python DDP: PythonDDP with support of SYNC and ASYNC reduction
- Added compare_ddp to measure the difference in forward and backward step

Kudos on Shen and Yi for the great idea.

Test Plan:

Test on DevGPUS with 2 CUDA devices.

$python compare_ddp.py

Python only DDP has slightly better (-1%) forward performance and slightly slower (2%-20%) backward performance.
This suggested that we need to keep C++ Core since the maximum latency increase can be 20%. See README.md for details.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29685364](https://our.internmc.facebook.com/intern/diff/D29685364)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

bowangbj commented:
Thanks Shen again for the great review. Fixed all the lint errors, landing ...

SciPioneer commented on benchmarks/distributed/ddp/compare/python_ddp.py:
```diff
@@ -0,0 +1,228 @@
+        self.async_reduction = async_reduction
+
+        # Holds all_reduce handles, used when async_reduction is True
+        self.async_handles = set()
```
Can we try using list instead of a set here? When the bucket size is large, the insertion and scanning of a hash set can be slower than a flat vector.

Instead, use `[None] * bucket_size` should be equivalent to a vector + reserve the size first.

SciPioneer commented on benchmarks/distributed/ddp/compare/python_ddp.py:
```diff
@@ -0,0 +1,228 @@
+        total_param = 0
+        for param in self.module.parameters():
+            total_param += 1
+            assert param.requires_grad, "param.requires_grad must be True"
```
Do we have a lot of parameters/layers? If so, `assert` here can slow down the loop, as it requires a sync between host and device.

bowangbj replied:
there are 1M params. Good idea -- we can introduce debug level to toggle it.

SciPioneer commented on benchmarks/distributed/ddp/compare/python_ddp.py:
```diff
@@ -0,0 +1,228 @@
+        """
+        # Validate bucket and offset are set.
+        bucket = self.param_to_bucket.get(param)
+        assert bucket is not None, "Failed to find bucket for param"
```
The `assert` in this method is also called in a loop.

bowangbj replied:
> @bowangbj Did you get a chance to collect the traces to compare the trainings using DDP core and Python Async DDP?

Thanks Yi for all the great comments. Sorry for the long delay. Finally I landed in PyTorch team after bootcamp :) 
Will follow up and dive into more details as part of my onboarding plan in Aug (thanks to Shen, Yi, Boris and Nikhil for the great proposal). 

Stay tuned!

SciPioneer commented on benchmarks/distributed/ddp/compare/python_ddp.py:
```diff
@@ -0,0 +1,228 @@
+
+        first_param = next(self.module.parameters())
+        for bucket in buckets:
+            bucket.buffer = first_param.new(bucket.total_elements)
```
`torch.Tensor.new` is deprecated, and it is likely to copy the data.

It will more efficient to use `new_empty`. See: https://pytorch.org/docs/stable/generated/torch.Tensor.new_empty.html

bowangbj replied:
Good idea, will address it in follow up CLs.

SciPioneer commented on benchmarks/distributed/ddp/compare/python_ddp.py:
```diff
@@ -0,0 +1,228 @@
+        if self.async_reduction:
+            for handle in self.async_handles:
+                handle.wait()
+            self.async_handles.clear()
```
If `async_handles` is a list of a fixed length, instead of a set, here we can just use `torch.futures.wait_all(async_handles)`, although I don't think it will make a difference in performance.

SciPioneer commented:
@bowangbj  Did you get a chance to collect the traces to compare the trainings using DDP core and Python Async DDP?
